tests/integration/snowflake/test_snowflake_adapter.py

```diff
@@ -0,0 +1,347 @@
+        """)
+
+    assert len(sf_adapter._safe_query(sample)) > 0
+
+
+def test_analyze_wrap_statement(sf_adapter):
+    DATABASE, SCHEMA, NAME, TABLE = "SNOWSHU_DEVELOPMENT", "POLYMORPHIC_DATA", "PARENT_TABLE", "PARENT_TABLE"
+    relation = Relation(database=DATABASE,
+                        schema=SCHEMA,
+                        name=NAME,
+                        materialization=TABLE,
+                        attributes=[])
+    sql = f'SELECT * from {DATABASE}.{SCHEMA}.{TABLE}'
+    statement = sf_adapter.analyze_wrap_statement(sql, relation)
+
+    assert query_equalize(statement) == query_equalize(f"""
+        WITH
+            {relation.scoped_cte('SNOWSHU_COUNT_POPULATION')} AS (
+        SELECT
+            COUNT(*) AS population_size
+        FROM
+            {sf_adapter.quoted_dot_notation(relation)}
+        )
+        ,{relation.scoped_cte('SNOWSHU_CORE_SAMPLE')} AS (
+        {sql}
+        )
+        ,{relation.scoped_cte('SNOWSHU_CORE_SAMPLE_COUNT')} AS (
+        SELECT
+            COUNT(*) AS sample_size
+        FROM
+            {relation.scoped_cte('SNOWSHU_CORE_SAMPLE')}
+        )
+        SELECT
+            s.sample_size AS sample_size
+            ,p.population_size AS population_size
+        FROM
+            {relation.scoped_cte('SNOWSHU_CORE_SAMPLE_COUNT')} s
+        INNER JOIN
+            {relation.scoped_cte('SNOWSHU_COUNT_POPULATION')} p
+        ON
+            1=1
+        LIMIT 1
+        """)
+
+
+def test_predicate_constraint_statement(sf_adapter):
+    DATABASE, SCHEMA, TABLE = "SNOWSHU_DEVELOPMENT", "POLYMORPHIC_DATA", "PARENT_TABLE"
+    LOCAL_KEY, REMOTE_KEY = "ID", "CHILD_ID"
+    relation = Relation(database=DATABASE,
+                        schema=SCHEMA,
+                        name=TABLE,
+                        materialization=TABLE,
+                        attributes=[])
+    relation.core_query = f"""
+        SELECT
+            *
+        FROM 
+            {DATABASE}.{SCHEMA}.{TABLE}
+            SAMPLE BERNOULLI (10)
+        """
+    statement = sf_adapter.predicate_constraint_statement(relation, True, LOCAL_KEY, REMOTE_KEY)
+
+    assert query_equalize(statement) == query_equalize(f"""
+        {LOCAL_KEY} IN 
+            ( SELECT  
+                {REMOTE_KEY}
+            AS {LOCAL_KEY}
+            FROM (
+        SELECT
+            *
+        FROM 
+            {DATABASE}.{SCHEMA}.{TABLE}
+            SAMPLE BERNOULLI (10)
+        ))
+        """)
+
+
+def test_get_connection(sf_adapter):
+    conn_string = sf_adapter.get_connection()
+    DATABASE = sf_adapter.credentials.database
+    USER = sf_adapter.credentials.user
+    PASSWORD = sf_adapter.credentials.password
+    ACCOUNT = sf_adapter.credentials.account
+    ROLE = sf_adapter.credentials.role
+
+    assert conn_string.url.render_as_string(hide_password=False) == \
+           f'snowflake://{USER}:{PASSWORD.replace("@", urllib.parse.quote_plus("@"))}@{ACCOUNT}/{DATABASE}/?role={ROLE}'
```

Comment:
This test might also be worth moving to unit tests with the other connection string test. Unless I'm missing something about this that actually calls out to snowflake.